ZEngine/include/ZEngine/Rendering/Buffers/IndexBuffer.h

```diff
@@ -41,18 +56,131 @@ namespace ZEngine::Rendering::Buffers
+                VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_INDEX_BUFFER_BIT,
+                memory_properties,
+                VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+        }
+
+        VkBuffer       staging_buffer;
+        VkDeviceMemory staging_buffer_memory;
+        Helpers::CreateBuffer(
+            performant_device,
+            staging_buffer,
+            staging_buffer_memory,
+            static_cast<VkDeviceSize>(this->m_byte_size),
+            VK_BUFFER_USAGE_TRANSFER_SRC_BIT,
+            memory_properties,
+            VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
+
+        void* memory_region;
+        ZENGINE_VALIDATE_ASSERT(vkMapMemory(device_handle, staging_buffer_memory, 0, this->m_byte_size, 0, &memory_region) == VK_SUCCESS, "Failed to map the memory")
+        std::memcpy(memory_region, content.data(), this->m_byte_size);
+        vkUnmapMemory(device_handle, staging_buffer_memory);
+
+        Helpers::CopyBuffer(performant_device, staging_buffer, m_index_buffer, static_cast<VkDeviceSize>(this->m_byte_size));
+
+        vkDestroyBuffer(device_handle, staging_buffer, nullptr);
+        vkFreeMemory(device_handle, staging_buffer_memory, nullptr);
+    }
+
+    template <>
+    inline void IndexBuffer<std::vector<uint32_t>>::SetData(const std::vector<uint32_t>& content)
+    {
+        Hardwares::VulkanDevice& performant_device = Engine::GetVulkanInstance()->GetHighPerformantDevice();
+        auto                     device_handle     = performant_device.GetNativeDeviceHandle();
+        const auto&              memory_properties = performant_device.GetPhysicalDeviceMemoryProperties();
+
+        if (this->m_byte_size < (content.size() * sizeof(float)))
+        {
+            CleanUpMemory();
+            GraphicBuffer<std::vector<uint32_t>>::SetData(content);
+            Helpers::CreateBuffer(
+                performant_device,
+                m_index_buffer,
+                m_index_buffer_device_memory,
+                static_cast<VkDeviceSize>(this->m_byte_size),
+                VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_INDEX_BUFFER_BIT,
+                memory_properties,
+                VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+        }
+
+        VkBuffer       staging_buffer;
+        VkDeviceMemory staging_buffer_memory;
+        Helpers::CreateBuffer(
+            performant_device,
+            staging_buffer,
+            staging_buffer_memory,
+            static_cast<VkDeviceSize>(this->m_byte_size),
+            VK_BUFFER_USAGE_TRANSFER_SRC_BIT,
+            memory_properties,
+            VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
+
+        void* memory_region;
+        ZENGINE_VALIDATE_ASSERT(vkMapMemory(device_handle, staging_buffer_memory, 0, this->m_byte_size, 0, &memory_region) == VK_SUCCESS, "Failed to map the memory")
+        std::memcpy(memory_region, content.data(), this->m_byte_size);
```

Comment:
I ran into a crash here. It looks like content might be empty sometimes, so content.data() is nullptr. Same goes for the other variants probably.

Reply:
I think the VertexBuffer class exhibits the same issue.

Reply:
Interesting. 
can you help with a repro case?

Reply:
Fixed :) ... thanks